cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/AppWithMapReduceUsingMultipleInputs.java

```diff
@@ -0,0 +1,154 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.batch.dataset.input;
+
+import co.cask.cdap.api.ProgramLifecycle;
+import co.cask.cdap.api.app.AbstractApplication;
+import co.cask.cdap.api.data.batch.Input;
+import co.cask.cdap.api.dataset.lib.FileSetArguments;
+import co.cask.cdap.api.dataset.lib.FileSetProperties;
+import co.cask.cdap.api.mapreduce.AbstractMapReduce;
+import co.cask.cdap.api.mapreduce.MapReduceContext;
+import co.cask.cdap.api.mapreduce.MapReduceTaskContext;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.apache.hadoop.mapreduce.lib.input.FileSplit;
+import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
+import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * App used to test whether M/R can read and perform a join across multiple inputs.
+ */
+public class AppWithMapReduceUsingMultipleInputs extends AbstractApplication {
+
+  public static final String PURCHASES = "purchases";
+  public static final String CUSTOMERS = "customers";
+  public static final String OUTPUT_DATASET = "saturatedRecords";
+
+  @Override
+  public void configure() {
+    setName("AppWithMapReduceUsingMultipleInputs");
+    setDescription("Application with MapReduce job using multiple inputs");
+    createDataset(PURCHASES, "fileSet", FileSetProperties.builder()
+      .setInputFormat(TextInputFormat.class)
+      .build());
+    createDataset(CUSTOMERS, "fileSet", FileSetProperties.builder()
+      .setInputFormat(TextInputFormat.class)
+      .build());
+    createDataset(OUTPUT_DATASET, "fileSet", FileSetProperties.builder()
+      .setOutputFormat(TextOutputFormat.class)
+      .setOutputProperty(TextOutputFormat.SEPERATOR, " ")
+      .build());
+    addMapReduce(new ComputeSum());
+  }
+
+  /**
+   * Computes sum of a customer's spending, while also joining on a lookup table, to get more data.
+   */
+  public static final class ComputeSum extends AbstractMapReduce {
+
+    @Override
+    public void beforeSubmit(MapReduceContext context) throws Exception {
+      Map<String, String> inputArgs = new HashMap<>();
+      FileSetArguments.setInputPath(inputArgs, "inputFile");
+
+      context.addInput(Input.ofDataset(PURCHASES, inputArgs), FileMapper.class);
```

Comment:
Perhaps we should allow user to `alias` their `Input`.
So something like:
```
context.setInput(Input.ofDataset("myPFS", args).alias("firstPFS");
context.setInput(Input.ofDataset("myPFS", otherArgs).alias("secondPFS");
```
What do you think about this? This would require that users alias their stream and dataset, if they have the same name.